pixeltable/env.py

```diff
@@ -816,6 +818,91 @@ def __init_spacy(self) -> None:
         except Exception as exc:
             raise excs.Error(f'Failed to load spaCy model: {spacy_model}') from exc
 
+    def _cleanup(self) -> None:
+        """
+        Internal cleanup method that properly closes all resources and resets state.
+        This is called before destroying the singleton instance.
+        """
+        # Close database connections and sessions
+        if self._current_session is not None:
+            try:
+                self._current_session.close()
+            except Exception as e:
+                _logger.error(f'Error closing session: {e}')
+            self._current_session = None
+
+        if self._current_conn is not None:
+            try:
+                self._current_conn.close()
+            except Exception as e:
+                _logger.error(f'Error closing connection: {e}')
+            self._current_conn = None
+
+        # Stop HTTP server
+        if self._httpd is not None:
+            try:
+                self._httpd.shutdown()
+                self._httpd.server_close()
+            except Exception as e:
+                _logger.error(f'Error stopping HTTP server: {e}')
```

Comment:
Are these all really errors or should they be warnings?

Reply:
we can ignore these actually. This was added too defensively when we were switching from control plane db to tenant db (now we have a separate env and its own engine/connection/session for control plane, we can lower severity down)